### PR TITLE
Rewrite unaryNot EXPR

### PR DESCRIPTION
**Pull request by Aiee**
Rewrite ```UnaryNotExpr``` so that indexes can be used during ```indexScan```. Following rules are supported:
1. Colasspe nested UnaryNot expression
    - ```e.g. !!(expr)  =>  (expr)```
2. Rewrite the combination of logical expr and unary expr
    - ```e.g. !(a && b && c)  =>  (!a || !b || !c)```
3. Rewrite the combination of relational expr and unary expr
    - ```e.g. !(a > b)  =>  (a <= b)```

Depends on https://github.com/vesoft-inc/nebula-common/pull/481

- [x] Add support to TCK frame so that operator info in the execution plan can be compared in the CI test. https://github.com/vesoft-inc/nebula-graph/pull/850

### CODE REVIEW

File: src/visitor/RewriteUnaryNotExprVisitor.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+
+namespace nebula {
+namespace graph {
+
+void RewriteUnaryNotExprVisitor::visit(ConstantExpression *expr) {
+    expr_.reset(expr->clone().release());
+}
+
+void RewriteUnaryNotExprVisitor::visit(UnaryExpression *expr) {
+    if (isUnaryNotExpr(expr)) {
```
Comment (yixinglu) [inline src/visitor/RewriteUnaryNotExprVisitor.cpp:19]:
```cpp
auto operand = expr->operand();
if (isUnaryNotExpr(expr) && isUnaryNotExpr(operand)) {
    operand->operand()->accept(this);
    expr_ = operand->operand();
    return;
} 
operand->accept(this);
expr_ = expr;

```

File: src/visitor/RewriteUnaryNotExprVisitor.cpp [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+void RewriteUnaryNotExprVisitor::visit(UnaryExpression *expr) {
+    auto operand = expr->operand();
+    if (isUnaryNotExpr(expr)) {
+        if (isUnaryNotExpr(operand)) {
+            static_cast<UnaryExpression *>(operand)->operand()->accept(this);
+            // reduce nested unary expr
+            if (expr_ != static_cast<UnaryExpression *>(operand)->operand()) {
+                return;
+            }
+            expr_ = static_cast<UnaryExpression *>(operand)->operand();
```
Comment (yixinglu) [inline src/visitor/RewriteUnaryNotExprVisitor.cpp:27]: useless

File: src/visitor/RewriteUnaryNotExprVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+// reduce expression
+void RewriteUnaryNotExprVisitor::visit(ReduceExpression *expr) {
+    expr_ = expr;
+}
+
+Expression* RewriteUnaryNotExprVisitor::reduce(UnaryExpression *expr) {
+    auto reducedExpr = static_cast<UnaryExpression *>(expr->operand())->operand();
+    return reducedExpr;
+}
```
Comment (yixinglu) [inline src/visitor/RewriteUnaryNotExprVisitor.cpp:239]: mark

File: src/visitor/test/RewriteUnaryNotExprVisitorTest.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        RewriteUnaryNotExprVisitor visitor;
+        expr->accept(&visitor);
+        auto res = visitor.getExpr();
+        auto expected = pool.add(neExpr(constantExpr(5), constantExpr(10)));
+        ASSERT_EQ(*res, *expected) << res->toString() << " vs. " << expected->toString();
+    }
+}
+
+TEST_F(RewriteUnaryNotExprVisitorTest, TestMultipleUnaryNotExprLogicalRelExpr) {
+    // !!(5 == 10) AND !!(30 > 20)  =>  (5 == 10) AND (30 > 20)
```
Comment (yixinglu) [inline src/visitor/test/RewriteUnaryNotExprVisitorTest.cpp:69]: !( a != 1 & B & C)

!(A || B || C)

**Review by yixinglu — COMMENTED** [review]
Generally LGTM, Good job!

File: src/visitor/RewriteUnaryNotExprVisitor.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    void visit(ListComprehensionExpression *) override;
+    // reduce expression
+    void visit(ReduceExpression *expr) override;
+
+    void visitBinaryExpr(BinaryExpression *expr) override;
+
+private:
+    Expression *expr_;
+    ObjectPool *pool_;
+
```
Comment (yixinglu) [inline src/visitor/RewriteUnaryNotExprVisitor.h:86]: Place all fields at the tail of class.

File: src/visitor/test/RewriteUnaryNotExprVisitorTest.cpp [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    }
+}
+
+TEST_F(RewriteUnaryNotExprVisitorTest, TestRelExpr) {
+    // (5 == 10)  =>  (5 == 10)  no change should be made
+    {
+        auto expr = pool.add(eqExpr(constantExpr(5), constantExpr(10)));
+        RewriteUnaryNotExprVisitor visitor(&pool);
+        expr->accept(&visitor);
+        auto res = visitor.getExpr();
```
Comment (yixinglu) [inline src/visitor/test/RewriteUnaryNotExprVisitorTest.cpp:143]: you should check whether the pointers of expr and res are same.

**Review by yixinglu — APPROVED** [review]
LGTM

File: src/util/ExpressionUtils.cpp [hunk 7]
```diff
@@ -236,10 +243,89 @@ Status ExpressionUtils::checkAggExpr(const AggregateExpression* aggExpr) {
     return Status::OK();
 }
 
-bool ExpressionUtils::isEvaluableExpr(const Expression* expr) {
-    EvaluableExprVisitor visitor;
-    const_cast<Expression*>(expr)->accept(&visitor);
-    return visitor.ok();
+std::unique_ptr<Expression> ExpressionUtils::reverseRelExpr(Expression *expr) {
+    auto left = static_cast<RelationalExpression *>(expr)->left();
```
Comment (jievince) [inline src/util/ExpressionUtils.cpp:247]: dont' need DCHECK here?

File: src/visitor/RewriteUnaryNotExprVisitor.cpp [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+    if (isUnaryNotExpr(expr)) {
+        if (isUnaryNotExpr(operand)) {
+            static_cast<UnaryExpression *>(operand)->operand()->accept(this);
+            return;
+        } else if (ExpressionUtils::isRelExpr(operand)) {
+            expr_ = pool_->add(ExpressionUtils::reverseRelExpr(operand).release());
+            return;
+        } else if (ExpressionUtils::isLogicalExpr(operand)) {
+            auto reducedLogicalExpr =
+                pool_->add(ExpressionUtils::reverseLogicalExpr(operand).release());
```
Comment (jievince) [inline src/visitor/RewriteUnaryNotExprVisitor.cpp:31]: Why add to pool?

**Review by jievince — DISMISSED** [review]
Awesome!

File: src/util/ExpressionUtils.cpp (hunk 7 above)
Comment (Aiee) [reply src/util/ExpressionUtils.cpp:247]: You are right. Thanks for catching this.

**Review by jievince — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Review by jievince — DISMISSED** [review]

**Review by yixinglu — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Pull request merged**